### PR TITLE
Fix an issue and improve diagnostics for target-based dependencies

### PR DESCRIPTION
**Pull request by hartbit**
Added a missing diagnostic to check that explicit dependency names match the package name and improve the diagnostic for invalid product target dependency references.

### CODE REVIEW

File: Sources/PackageModel/Manifest.swift [hunk 1]
```diff
@@ -267,7 +267,7 @@ extension Manifest {
 
     /// Finds the package dependency referenced by the specified target dependency.
     /// - Returns: Returns `nil` if the dependency is a target dependency, if it is a product dependency but has no
-    /// package name (for tools versions less than 5.2), or if there were no dependencies with the provided name.
+    /// package name (for tools versions less than 5.3), or if there were no dependencies with the provided name.
```
Comment (aciidb0mb3r) [inline Sources/PackageModel/Manifest.swift]: This should say 5.2, right?

File: Tests/PackageGraphTests/PackageGraphTests.swift [hunk 2]
```diff
@@ -548,7 +548,87 @@ class PackageGraphTests: XCTestCase {
...
+                dependency 'BarLib' in target 'Foo' requires explicit declaration; reference the package in the target \
+                dependency with '.product(name: "BarLib", package: "Bar")'
+                """, behavior: .error, location: "'Foo' /Foo")
+            result.checkUnordered(diagnostic: """
+                dependency 'Biz' in target 'Foo' requires explicit declaration; provide the name of the package \
+                dependency with '.package(name: "Biz", url: "/BizPath", .exact("1.2.3"))'
+                """, behavior: .error, location: "'Foo' /Foo")
+            result.checkUnordered(diagnostic: """
+                dependency 'FizLib' in target 'Foo' requires explicit declaration; reference the package in the target \
+                dependency with '.product(name: "FizLib", package: "FizPath")' and provide the name of the package \
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/PackageGraphTests.swift]: Shouldn't this say "Fiz" and not "FizPath" in the package name?

Comment (aciidb0mb3r) [reply Tests/PackageGraphTests/PackageGraphTests.swift]: Also, the name label is not present in `.product`, i.e., it's `.product("productname", package: "...")`

File: Tests/PackageGraphTests/PackageGraphTests.swift [hunk 3]
```diff
@@ -818,6 +898,49 @@ class PackageGraphTests: XCTestCase {
...
+                    ],
+                    targets: [
+                        TargetDescription(name: "Baar"),
+                    ]),
+            ]
+        )
+
+        DiagnosticsEngineTester(diagnostics, ignoreNotes: true) { result in
+            result.check(
+                diagnostic: "package dependency 'Baar' does not match the package name 'Bar'",
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/PackageGraphTests.swift]: How about something like:

"The name declared ('Baar') for dependency `"/Bar"` doesn't match the actual package name `'Bar'`.

**Comment by aciidb0mb3r** [comment]
Thanks! This looks good, I left some minor comments.

File: Sources/PackageModel/Manifest.swift (hunk 1 above)
Comment (hartbit) [reply Sources/PackageModel/Manifest.swift]: Correct. My bad.

File: Tests/PackageGraphTests/PackageGraphTests.swift (hunk 2 above)
Comment (hartbit) [reply Tests/PackageGraphTests/PackageGraphTests.swift]: Thanks! Fixed.

File: Tests/PackageGraphTests/PackageGraphTests.swift (hunk 3 above)
Comment (hartbit) [reply Tests/PackageGraphTests/PackageGraphTests.swift]: Improved it, thanks!

**Comment by hartbit** [comment]
@swift-ci please smoke test

**Pull request merged**